autopilot/ipfilter_test.go

```diff
@@ -0,0 +1,155 @@
+	ipr.cache["example.com:1234"] = ipCacheEntry{created: time.Now(), subnets: []string{"a"}}
+	r.setNextErr(errIOTimeout)
+	if subnets, err := ipr.lookup("example.com:1234"); err != nil {
+		t.Fatal("unexpected error", err)
+	} else if len(subnets) != 1 || subnets[0] != "a" {
+		t.Fatal("unexpected subnets", subnets)
+	}
+
+	// test too many addresses - more than two
+	r.setAddr("example.com", []net.IPAddr{{}, {}, {}})
+	if _, err := ipr.lookup("example.com:1234"); !isErr(err, errTooManyAddresses) {
+		t.Fatal("unexpected error", err)
+	}
+
+	// test too many addresses - two of the same type
+	r.setAddr("example.com", []net.IPAddr{{IP: net.IPv4(1, 2, 3, 4)}, {IP: net.IPv4(1, 2, 3, 4)}})
+	if _, err := ipr.lookup("example.com:1234"); !isErr(err, errTooManyAddresses) {
+		t.Fatal("unexpected error", err)
+	}
+
+	// test invalid addresses
+	r.setAddr("example.com", []net.IPAddr{{IP: ipv4Localhost}, {IP: net.IP{127, 0, 0, 2}}})
+	if _, err := ipr.lookup("example.com:1234"); !isErr(err, errTooManyAddresses) {
+		t.Fatal("unexpected error", err)
+	}
+
+	// test valid addresses
+	r.setAddr("example.com", []net.IPAddr{{IP: ipv4Localhost}, {IP: ipv6Localhost}})
+	if subnets, err := ipr.lookup("example.com:1234"); err != nil {
+		t.Fatal("unexpected error", err)
+	} else if len(subnets) != 2 || subnets[0] != "127.0.0.0/24" || subnets[1] != "::/32" {
+		t.Fatal("unexpected subnets", subnets)
+	}
+}
+
+func TestIPFilter(t *testing.T) {
+	r := newTestResolver()
+	r.setAddr("host1.com", []net.IPAddr{{IP: net.IP{192, 168, 0, 1}}})
+	r.setAddr("host2.com", []net.IPAddr{{IP: net.IP{192, 168, 1, 1}}})
+	r.setAddr("host3.com", []net.IPAddr{{IP: net.IP{192, 168, 2, 1}}})
+	ipf := newTestIPFilter(r)
+
+	// add 3 hosts - unique IPs
+	r1 := ipf.IsRedundantIP("host1.com:1234", types.PublicKey{1})
+	r2 := ipf.IsRedundantIP("host2.com:1234", types.PublicKey{2})
+	r3 := ipf.IsRedundantIP("host3.com:1234", types.PublicKey{3})
+	if r1 || r2 || r3 {
+		t.Fatal("unexpected result", r1, r2, r3)
+	}
+
+	// try add 4th host - redundant IP
+	r.setAddr("host4.com", []net.IPAddr{{IP: net.IP{192, 168, 0, 12}}})
+	if redundant := ipf.IsRedundantIP("host4.com:1234", types.PublicKey{4}); !redundant {
+		t.Fatal("unexpected result", redundant)
+	}
+
+	// add 4th host - unique IP - 2 subnets
+	r.setAddr("host4.com", []net.IPAddr{{IP: net.IP{192, 168, 3, 1}}, {IP: net.ParseIP("2001:0db8:85a3::8a2e:0370:7334")}})
+	if redundant := ipf.IsRedundantIP("host4.com:1234", types.PublicKey{4}); redundant {
+		t.Fatal("unexpected result", redundant)
+	}
+
+	// try add 5th host - redundant IP based on the IPv6 subnet from host4
+	r.setAddr("host5.com", []net.IPAddr{{IP: net.ParseIP("2001:0db8:85b3::8a2e:0370:7335")}})
+	if redundant := ipf.IsRedundantIP("host5.com:1234", types.PublicKey{5}); !redundant {
+		t.Fatal("unexpected result", redundant)
+	}
+
+	// add 5th host - unique IP
+	r.setAddr("host5.com", []net.IPAddr{{IP: net.ParseIP("2001:0db9:85b3::8a2e:0370:7335")}})
```

Comment:
the `ipv6FilterRange` surprised me but I remember there was a reason for using "only" 32 bits, do you remember @ChrisSchinnerl ?

Reply:
I compared the subnets to the known IP ranges that hosting providers own.